doc/dns-resovler.md

```diff
@@ -0,0 +1,58 @@
+# DNS Resolver 
+
+APIcast ships with own DNS resolver implemented in Lua.
+
+## Motivation
+
+Current nginx resolver has some issues like that it does not support multiple DNS servers properly.
+For example on OpenShift there are two DNS servers and one search path. Nginx is not able to properly resolve internal DNS names.
+
+Also we like to use nginx facilities for load balancing and proxying. That means we have to use `balancer_by_lua` directive which can set the load balancer. But it has to set it to the IP address, not the DNS name. So we have to resolve that DNS name into IP address and there is no API exposed in nginx to do that.
+
+## Requirements
+
+Our DNS resolver should meet following criteria:
+
+* caching for duration of TTL
+* resolve A and CNAME records
+* using stale cache when TTL expires and cache can't be updated
+* many parallel requests to one worker result in one query
+* connect to multiple DNS servers in parallel
+* working search scope
+
+## Scenarios
+
+### Low traffic
+
+A request is made to APIcast and DNS cache is updated. Then there is no request for TTL of that record and another request is made after that record expires. Should APIcast try to use stale cache or hold the request? For how long?
```

Comment:
Yes, the idea is to hide resolving latencies as much as possible. Also, when the oob refreshing cannot refresh fast enough to avoid a window of time in which the TTL is expired but the resolving has not finished, deciding to return stale data back is probably good enough (as opposed to waiting on the refreshing to finish), and is consistent with the behaviour that would occur if the refresh never actually finished or didn't do it in a reasonable time.